### ISSUE TITLE
Redesign equations

### ISSUE THREAD

**Issue by tkarna** [thetisproject/thetis#26]
# Redesign equations

The `Equation` class as it is implemented now is too rigid for developing custom time integrators. We need a more flexible API where the user can choose which terms are treated implicitly/explicitly/differently.

## Equations and terms

Equation class defines all the terms in the equation. It takes a bunch of options and the function space where the solution belongs:

```python
function_space = solver.function_spaces.P1DG
options = {'use_grad_div_viscosity': True}
eq = MyEquation(function_space, **options)

class MyEquation(Equation):
    def __init__(self, function_space, ...)
        self.function_space = function_space
        self.test = TestFunction(self.function_space)
        ...
```

Options are typically booleans like above.

The terms are implemented as methods that return the corresponding form, including interface terms and boundary conditions

```python
class MyEquation(Equation):
    ...
    def viscosity(self, fields, bnd_conditions):
        """viscosity operator with interior penalty terms."""
        grad_div = self.options.use_grad_div_viscosity
        f = 0
        f += ... self.test * dx
        return f
```

Sign convention: all terms are assumed to be on the left hand side of the equation A + term = 0.

The terms depend on a `fields` dictionary

```python
fields={'uv_3d': solver.fields.uv_3d,
        'elev_3d': solver.fields.elev_3d,
        'viscosity': None}
```

Fields should be `Coefficient`s, e.g. `Function` or `Constant` instances, or UFL expressions(?). Some fields are optional, e.g. if `viscosity` is omitted (or it is `None`) the corresponding terms will be ignored. (Alternatively we could have `option.use_viscosity`.) Some fields are required and should be asserted.

The equation object knows all the terms that it implements

```python
class MyEquation(Equation):
    def __init__(...):
        ...
        # add some python magic to do this
        self.terms = {}
        self.terms['viscosity'] = self.viscosity
        self.terms['elevation_gradient'] = self.elevation_gradient
```

To differentiate how the terms should be treated, the user can add labels:

```python
eq.set_label('implicit', ('viscosity', 'elevation_gradient'))
eq.set_label('explicit', 'advection')
eq.set_label('source', 'wind_stress')
```

The equation class provides a method for evaluating the forms

```
f_implicit = eq.get_form('implicit', fields)
# returns eq.viscosity(fields) + eq.elevation_gradient(fields)

f_explicit = eq.get_form('explicit', fields)
# returns eq.advection(fields)
```

## Time integrator

Time integrator then takes the equation, fields and labels as an argument 

```python
labels = {'implicit': ('viscosity', 'elevation_gradient'),
          'explicit': 'advection'}
ti = MyTimeIntegrator(eq, fields, labels, bnd_conditions)
```

The labels depend on the time integrator, e.g. backward Euler only has implicit terms.
Using the `Equation.get_form` method the time integrator can then construct the solvers it needs:

```python
class MyTimeIntegrator(Equation):
    def __init__(self, equation, fields, labels, bnd_conditions):
        self.equation = equation
        ...
        mass_term = ...
        f_k1 = mass_term + self.dt*self.equation.get_form('implicit', fields_k1)
        self.problem_k1 = NonlinearVariationalProblem(f_k1, self.solution_k1)
        self.solver_k1 = NonlinearVariationalSolver(self.problem_k1)

    def set_timestep(self, dt):
        self.dt.assign(dt)

    def advance(self, time, update_forcings=None):
        ...
```
        
NOTE: Currently the equation object knows the fields it uses and also the solution function. These however may depend on the time integrator so it's better to define them here.

## Boundary conditions

Boundary conditions are defined in a dictionary that define which fields should be replaced at the boundary (for weak boundary conditions). There can be additional boundary specific options.

```python
ocean_bnd_id = 1
river_bnd_id = 2
ocean_uv = Constant((1, 0, 0))
river_uv = Constant((0.05, 0, 0))
ocean_funcs = {'uv_3d': ocean_uv}
river_funcs = {'uv_3d': river_uv, 'clamped': True}
bnd_advection = {ocean_bnd_id: ocean_funcs,
                 river_bnd_id: river_funcs}
    
# apply to term
f_adv = eq.advection(fields, bnd_advection)

# or time integrator
bcs = {'advection': bnd_advection}
ti = MyTimeIntegrator(eq, fields, labels, bnd_conditions)
```

It should be possible to support Dirichlet boundary conditions in a similar fashion,
```python
ocean_funcs = DirichletBC(...)
```
this just will be passed to the solver instead of the term.

## Naming convention

For the sake of clarity the field names should follow `field_metadata` (defined in [field_defs.py](../blob/master/thetis/field_defs.py) ) as closely as possible. Likewise, options should be similar to [options.py](../blob/master/thetis/options.py) where applicable.

**Comment by tkarna** [comment]
Additionally we might be able to leverage dolfin-adjoint timestepping library for defining the time integrators.

**Comment by stephankramer** [comment]
As discussed offline, we should probably consider time integration schemes that lie inbetween the extremes of completely explicit and fully nonlinear implicit, to support things like IMEX schemes, subcycling of advection, and linearised implicit time schemes.

As an example, the current pressure gradient term in the depth-integrated equations looks like:

    -g*head*nabla_div(u_test) + (avg(head) + 0.5*sqrt(avg(total_h)/g)*jump(uv,n)) * jump(u_test,n)*dS

A naive, fully implicit time integration would create a non-linearity in the Riemann jump term as both uv and total_h=h+head are solved for. This leads to horrendous terms in the Jacobian that are often poorly behaved. For instance with a quadratic drag term:

    c_d*sqrt(dot(u, u))*u

the Jacobian isn't defined at u=0, and you actually need to use a non-zero initial guess for the non-linear solve to converge.

In my mind, for our purposes in all cases where implicit time integration is needed there is an obvious choice to get a linear implicit scheme. I would actually go as far to say that for our purposes, non-linear solves are really not that useful at all (except if you go to the extreme of steady state solves with artificially high viscosities which we sometimes do in tidal turbine optimisations). You can go sort of half-way with Picard iterations, where you update the factors that were treated explicitly in the linearisation in a non-linear iteration - but even that I think is superfluous in a  lot of cases. Just going for a purely linear implcit time integration only (i.e do just one Picard iteration) gives you a mathematically well-defined and stable scheme.

In terms of redesigning the code to support this, we would need all terms to take both a new and old (or lagged) velocity and pressure as argument:

    def pressure_gradient(head, uv, head_old, uv_old, ...)
       total_h_old = depth  + head_old
       return -g*head*nabla_div(u_test) + (avg(head) + 0.5*sqrt(avg(total_h_old)/g)*jump(uv,n)) * jump(u_test,n)*dS

Completely explicit, linear implicit and full non-linear implicit can then be supported by sticking in the right functions.

**Comment by tkarna** [comment]
All right, so how about something like this:

### Terms

Terms are objects that take some options and provide a method for getting the form:

```python
pg = PressureGradientTerm(constant_depth=True)
f = pg.get_residual(solution, solution_old, fields, field_old)
```

Above `fields` is a dictionary of fields other than the solution.

For implicit terms there may also be a routine for evaluating the Jacobian

```python
f = pg.get_jacobian(solution, solution_old, fields, field_old)
```

### Equations

```python
eq = MomentumEquation(quadratic_bottom_drag=True)
eq.label_term(('PressureGradientTerm', 'QuadraticFrictionTerm'), 'implicit')
eq.label_term('AdvectionTerm', 'explicit')
f = eq.get_residual('explicit', solution, solution_old, fields, fields_old)
g = eq.get_residual('implicit', solution, solution_old, fields, fields_old)
j = eq.get_jacobian('implicit', solution, solution_old, fields, fields_old)
```

Equation object has a list of default terms that it always evaluates. If some field required by a term is missing, the term is dropped. Some terms are also included/excluded depending on the options (e.g. linear bottom drag).

The terms can be labeled in 4 categories:

 - source: terms that do not depend on solution  
 - explicit: terms that depend on old solution  
 - implicit: terms that depend on old solution and are _linear_ w.r.t. solution
 - nonlinear: terms that are nonlinear w.r.t. solution

By default each term belongs to some category (e.g. source or explicit).    

User can also create custom equations easily

```python
eq = Equation()
eq.add_term(PressureGradientTerm, 'explicit')
eq.add_term(QuadraticBottomFrictionTerm, 'explicit')
```

- There should also be a way to omit a term (useful at least for debugging)

### Time integrators

Time intergrators take the equation, solution function, other fields and time step. They allocate all time lagged functions they might need.

```python
solution = Function(...)
dt = 1.0
ti = timeintegrator.CrankNicholson(eq, solution, fields, dt)
ti.advance(t, dt, update_forcings=None)
```

Generic time integrators will evaluate the different terms like this
- source:  
`f = term.get_residual(None, None, fields_old, fields_old)`
- explicit:  
`f = term.get_residual(solution_old, solution_old, fields_old, fields_old)`
- implicit:  
`f = term.get_residual(solution, solution_old, fields, fields_old)`
- nonlinear:  
`f = term.get_residual(solution, solution, fields, fields)`

Custom time integrators can deviate from this. For example some fields may need to be lagged and some not.

### Boundary conditions

Like in the first draft boundary conditions are defined in a dictionary, and passed to the time integrator, and all the way to the terms who know how to treat them.

**Comment by tkarna** [comment]
Implemented in #35
